scripts/Stretch.py

```diff
@@ -0,0 +1,419 @@
+        iflush = 0
+        for i, step in enumerate(layer_steps):
+            if i == 0:
+                current_e = step.step_e
+            if current_e == step.step_e:
+                # No extrusion since the previous step, so it is a travel move
+                # Let process steps accumulated into orig_steps,
+                # which are a sequence of continuous extrusion
+                modif_seq = np.copy(orig_seq)
+                if len(orig_seq) >= 2:
+                    self.workOnSequence(orig_seq, modif_seq)
+                layergcode = self.generate(layer_steps, iflush, i, modif_seq, layergcode)
+                iflush = i
+                orig_seq = np.empty((0, 2))
+            if step.step == 0 or step.step == 1:
+                orig_seq = np.concatenate([orig_seq, np.array([[step.step_x, step.step_y]])])
+            current_e = step.step_e
+        if len(orig_seq):
+            modif_seq = np.copy(orig_seq)
+        if len(orig_seq) >= 2:
+            self.workOnSequence(orig_seq, modif_seq)
+        layergcode = self.generate(layer_steps, iflush, len(layer_steps), modif_seq, layergcode)
+        return layergcode
+
+    def stepToGcode(self, onestep):
+        """
+        Converts a step into G-Code
+        For each of the X, Y, Z, E and F parameter,
+        the parameter is written only if its value changed since the
+        previous g-code step.
+        """
+        sout = ""
+        if onestep.step_f != self.outpos.step_f:
+            self.outpos.step_f = onestep.step_f
+            sout += " F{:.0f}".format(self.outpos.step_f).rstrip(".")
+        if (onestep.step_x != self.outpos.step_x or onestep.step_y != self.outpos.step_y
+                or onestep.step_z != self.outpos.step_z):
+            assert onestep.step_x >= -1000 and onestep.step_x < 1000 # If this assertion fails,
+                                                           # something went really wrong !
+            self.outpos.step_x = onestep.step_x
+            sout += " X{:.3f}".format(self.outpos.step_x).rstrip("0").rstrip(".")
+            assert onestep.step_y >= -1000 and onestep.step_y < 1000 # If this assertion fails,
+                                                           # something went really wrong !
+            self.outpos.step_y = onestep.step_y
+            sout += " Y{:.3f}".format(self.outpos.step_y).rstrip("0").rstrip(".")
+        if onestep.step_z != self.outpos.step_z:
+            self.outpos.step_z = onestep.step_z
+            sout += " Z{:.3f}".format(self.outpos.step_z).rstrip("0").rstrip(".")
+        if onestep.step_e != self.outpos.step_e:
+            self.outpos.step_e = onestep.step_e
+            sout += " E{:.5f}".format(self.outpos.step_e).rstrip("0").rstrip(".")
+        return sout
+
+    def generate(self, layer_steps, i, iend, orig_seq, layergcode):
+        """
+        Appends g-code lines to the plugin's returned string
+        """
+        ipos = 0
+        while i < iend:
```

Comment:
Well... My code is not easily readable... I will try to change again variables names here.
The i variable is a parameter, it is the starting index for the loop, and iend is the end index. The objective is to advance from the last output position (iflush in the calling function) to the position iend.
Your first proposition would restart from the beginning of the layer, and your second proposition would start from the beginning of the layer, and up to the end of the layer